apisix/kubernetes.lua

```diff
@@ -0,0 +1,330 @@
+            headers = {
+                ["Host"] = _apiserver.host .. ":" .. _apiserver.port,
+                ["Authorization"] = "Bearer " .. _apiserver.token,
+                ["Accept"] = "application/json",
+                ["Connection"] = "keep-alive"
+            }
+        })
+
+        core.log.info("--raw=" .. informer.path .. "?" .. informer:watch_query())
+
+        if err then
+            return false, "RequestError", err
+        end
+
+        if res.status ~= 200 then
+            return false, res.reason, res:read_body() or ""
+        end
+
+        local remainder_body = ""
+        local body
+        local reader = res.body_reader
+        local gmatch_iterator
+        local captures
+        local captured_size = 0
+
+        while true do
+            body, err = reader()
+            if err then
+                return false, "ReadBodyError", err
+            end
+
+            if not body then
+                break
+            end
+
+            if #remainder_body ~= 0 then
+                body = remainder_body .. body
+            end
+
+            gmatch_iterator, err = ngx.re.gmatch(body, "{\"type\":.*}\n", "jao")
+            if not gmatch_iterator then
+                return false, "GmatchError", err
+            end
+
+            captures, err = gmatch_iterator()
+
+            if err then
+                return false, "GmatchError", err
+            end
+
+            if not captures then
+                break
+            end
+
+            captured_size = captured_size + #captures[0]
+            local v, _ = core.json.decode(captures[0])
+
+            if not v or not v.object then
+                return false, "UnexpectedBody", captures[0]
+            end
+
+            local type = v.type
+            if type == _constants.ErrorEvent then
+                if v.object.code == _constants.ErrorGone then
+                    return true, "Success", nil
+                end
+                return false, "UnexpectedBody", captures[0]
+            end
+
+            local object = v.object
+            informer.version = object.metadata.resourceVersion
+
+            if type == _constants.AddedEvent then
+                if informer.on_added ~= nil then
+                    informer:on_added(object, _constants.WatchDrive)
+                end
+            elseif type == _constants.DeletedEvent then
+                if informer.on_deleted ~= nil then
+                    informer:on_deleted(object)
+                end
+            elseif type == _constants.ModifiedEvent then
+                if informer.on_modified ~= nil then
+                    informer:on_modified(object)
+                end
+                -- elseif type == _constants.BookmarkEvent then
+                --    do nothing
+            end
+
+            if captured_size == #body then
+                remainder_body = ""
+            elseif captured_size == 0 then
+                remainder_body = body
+            else
```

Comment:
There will be bugs in some scenarios. Imagine: when an event is divided into two chunks, the first half of the first cycle is kept in the remainder_body. Since capture_size is not reset at this time, after the second loop comes in, the capture_size calculation is wrong.